### PR TITLE
App uplinks probing and failover to the best uplink

### PR DESCRIPTION
**Pull request by naiming-zededa**
this is the patch for periodical ping the link nexthop and remote host, and pick the best link available with link classes, and update the routes/ACL in app tables.

### CODE REVIEW

**Comment by rvs** [comment]
Hey @naiming-zededa this PR needs to be restructured -- it is very difficult to review it given the enormous amount of merge commits. Let me know if you need any help with git rebase.

**Comment by naiming-zededa** [comment]
@rvs I only know how to rebase before i push to remove; but I need help on how to rebase this in the current state.

**Comment by rvs** [comment]
Hey @naiming-zededa it seems I have fixed dependencies issue. We're now back to a legit yetus complain https://3244-182198941-gh.circle-artifacts.com/0/tmp/yetus-out/diff-patch-revive.txt plus one unsigned commit in the middle. Let's huddle tomorrow and fix it.

**Comment by naiming-zededa** [comment]
@rvs Cool, Thanks Roman, i'll get the real yetus complaint fixed first.

File: pkg/pillar/cmd/zedrouter/probe.go [hunk 1]
```diff
@@ -0,0 +1,538 @@
...
+	if aa == nil { // no information
+		log.Infof("getIntfClassIOBundle: aa is nil")
+		return types.ClassETHER
+	}
+	ioBundle := aa.LookupIoBundleNet(port.IfName)
+	if ioBundle == nil {
+		log.Infof("getIntfClassIOBundle: iobundle is nil")
+		return types.ClassETHER
+	}
+	switch ioBundle.Type {
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedrouter/probe.go]: I don't think we should define classes but instead add a numeric priority in the API. For now you can use the freeUplink which is in DeviceNetworkStatus; later we'll add  the numeric priority end to end.

File: pkg/pillar/types/zedroutertypes.go [hunk 2]
```diff
@@ -782,6 +782,53 @@ const (
 	MST_LAST = 255
 )
 
+// IntfClass - enum definition for device Interface Class
+type IntfClass uint8
+
+// enum ddfinition for interface classes
```
Comment (eriknordmark) [inline pkg/pillar/types/zedroutertypes.go]: I don't think we should define classes but instead add a numeric priority in the API. For now you can use the freeUplink which is in DeviceNetworkStatus; later we'll add  the numeric priority end to end.

Comment (naiming-zededa) [reply pkg/pillar/types/zedroutertypes.go]: Ok, will modify that part of the patch.

**Comment by naiming-zededa** [comment]
Updated with changes from Class based to Free/Non-Free, and add a signal when first assigned UP link, and make the convergence parameters a little faster.
Still have two changes need to do:
- not to publish the NI status per probing and make stats local, that is too much noise in the system
- handle the dnsmasq comment

**Comment by eriknordmark** [comment]
Can you please rebase/squash so it doesn't have hundereds of modified files?

What I do is:
save a copy of the content of the files I modified (to check the content didn't get messed up)
git fetch upstream master
git rebase -i upstream/master
# Resolve any issues with applying the patches; squash where it makes sense.
git push -f

**Comment by gkodali-zededa** [comment]
> Once piece which is missing is to have the dnsmasq serving a given network instance use an upstream DNS server which is on the chosen/current port.
> 
> There is a function (not yet exported) in the devicenetwork called ifnameToResolvConf() which you can export and use for this purpose. There is a config setting for dnsmasq to have it not use /etc/resolv.conf and one to specify which file to use instead/in addition.

I'll look into this @nordmark

**Comment by gkodali-zededa** [comment]
> > Once piece which is missing is to have the dnsmasq serving a given network instance use an upstream DNS server which is on the chosen/current port.
> > There is a function (not yet exported) in the devicenetwork called ifnameToResolvConf() which you can export and use for this purpose. There is a config setting for dnsmasq to have it not use /etc/resolv.conf and one to specify which file to use instead/in addition.
> 
> I'll look into this @nordmark

@nordmark I have added code for this and committed.

**Comment by gkodali-zededa** [comment]
@naiming-zededa Should we just close this PR?

**Comment by naiming-zededa** [comment]
close this PR, moved to https://github.com/lf-edge/eve/pull/291

**Pull request closed without merging**